Harmonic notch filter implementation

andyp1per opened the pull request:
New HarmonicNotchFilter that delegates updates to a bank of NotchFilters located at an rpm frequency and harmonics. Center frequency can be updated dynamically. Notch parameters are configurable, including the number of harmonics to filter on. Updates to the filter parameters are optimized across the notch bank.

Example configuration is as follows:

INS_HNTCH_ATT    15.000000
INS_HNTCH_FREQ 80.000000
INS_HNTCH_BW     20.000000
INS_HNTCH_ENABLE 1.000000
INS_HNTCH_HMNCS  1.000000

khancyr commented on libraries/Filter/NotchFilter.cpp:
```diff
@@ -21,10 +21,16 @@
+}
+
+template <class T>
+void NotchFilter<T>::_init(float sample_freq_hz, float center_freq_hz, float A, float Q)
```
this is confusing

andyp1per replied:
@khancyr Would internal_init() be better?

andyp1per replied:
I changed to internal_init()

khancyr commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,158 @@
+    // @DisplayName: Harmonics
+    // @Description: Number of harmonic frequencies to add to the filter. This option takes effect on the next reboot.
+    // @Range: 0 2
+    // @User: Advanced
```
// @RebootRequired: True

andyp1per replied:
Thanks! Fixed

lthall commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,159 @@
+
+    // @Param: ENABLE
+    // @DisplayName: Enable
+    // @Description: Enable dynamic notch filter
```
We need to change "dynamic" to "harmonic" in the comments.

andyp1per replied:
Thanks fixed

lthall commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,159 @@
+    float octaves = log2f(_center_freq_hz  / (_center_freq_hz - bandwidth_hz/2.0f)) * 2.0f;
+    float A = powf(10, -attenuation_dB/40.0f);
+    float Q = sqrtf(powf(2, octaves)) / (powf(2,octaves) - 1.0f);
+
```
We should not be reusing this code. We could make it a seperate function in the Notch filter and return those values so that way the same code is used in all three places.

andyp1per replied:
Normally I am keen not to bounce between virtuals, but in this case it makes it cleaner so have fixed that way.

lthall commented:
This is one of the most significant filter changes we have had for a long time!

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,159 @@
+    float octaves = log2f(center_freq_hz  / (center_freq_hz - bandwidth_hz/2.0f)) * 2.0f;
+    float A = powf(10, -attenuation_dB/40.0f);
+    float Q = sqrtf(powf(2, octaves)) / (powf(2,octaves) - 1.0f);
+
```
I see we can do that with A, but doesn't Q vary with the center frequency?

lthall replied:
(2 * center_freq_hz  / (2 *center_freq_hz - 2 *bandwidth_hz/2.0f))
=
(2 * center_freq_hz  /( 2 *(center_freq_hz - bandwidth_hz/2.0f)))
=
(center_freq_hz  / (center_freq_hz - bandwidth_hz/2.0f))

As both bandwidth and center frequency should scale by the harmonic, the Q stays constant.

lthall replied:
Sorry, I don't know why that went large and bold.

andyp1per replied:
Right, constant across the harmonics for the same centre frequency, but if we change the centre frequency itself (and thus change it across all harmonics), Q will change?

lthall replied:
no, it won't. The Q should stay the same for all harmonics and for the full range of centre frequency.

When we move the centre frequency up and down we want the ratio of centre frequency and bandwidth to remain constant.

andyp1per replied:
So let me rephrase what I think you are saying - if we change the centre frequency we want to change the bandwidth as well such that Q remains constant?

lthall replied:
Sorry that was not clear. Yes you are correct. 

If we are using this dynamically we would want to keep the Q constant (ratio of centre frequency and bandwidth constant).

andyp1per replied:
Ok thanks. I have undone my previous change and refatored on that basis.

lthall replied:
Perfect!!

The only thing is I think our standard is to put an underscore before variables like this so I believe it should be _A and _Q.

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,158 @@
+        return;
+    }
+
+    for (int i=0; i<harmonics+1; i++) {
```
small thing but we don't normally use "int".. more often "int16_t" or "int8_t"

andyp1per replied:
Thanks, I'll clean this up with the variable naming Leonard wanted

andyp1per replied:
Ok fixed

khancyr commented on libraries/Filter/NotchFilter.h:
```diff
@@ -29,12 +29,24 @@
 
+    // calculate attenuation and quality from provided center frequency and bandwidth
+    static void calculate_A_and_Q(float center_freq_hz, float bandwidth_hz, float attenuation_dB, float& A, float& Q) {
+        const float octaves = log2f(center_freq_hz / (center_freq_hz - bandwidth_hz / 2.0f)) * 2.0f;
```
shouldn't we check that (center_freq_hz - bandwidth_hz / 2.0f != 0) ?

andyp1per replied:
I've done that in the caller:

// adjust the center frequency to be in the allowable range
    center_freq_hz = constrain_float(center_freq_hz, bandwidth_hz * 0.52f, sample_freq_hz * 0.48f);

This was actually a bug in the original notch, pretty sure you can crash the FC if you select some inappropriate values.

andyp1per commented:
@OXINARF Leonard and I are all done - ready for review.

jschall commented:
Does this dynamically change the frequency being notched based on RPM feedback from ESCs? I took a brief look and it doesn't seem like it.

andyp1per commented:
> Does this dynamically change the frequency being notched based on RPM feedback from ESCs? I took a brief look and it doesn't seem like it.

Not yet - but it is a crucial building block towards that

lvale commented:
If we had a FFT running this would be perfect.

andyp1per commented:
> If we had a FFT running this would be perfect.

Leonard and I have a working prototype that uses throttle output to modify the center frequency. FFT is more complicated, but on the radar. Ultimately bi-directional dshot will provide the most accurate information.

OXINARF commented:
@andyp1per I've just finished my review, but there is so much changes between commits (including adding and then removing stuff) that I have many outdated comments.

Is this finished? If it is, please squash the commits, then I'll check my comments against it and post the review - please write a comment when you've done it so that I'm aware.

andyp1per commented:
> Is this finished? If it is, please squash the commits, then I'll check my comments against it and post the review - please write a comment when you've done it so that I'm aware.
@OXINARF my apologies. Yes it is finished. I found a small structural smell yesterday when doing test flying which I have resolved. Will squash now.

andyp1per commented:
@OXINARF squash is done and I have pushed

OXINARF commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -211,6 +214,15 @@ class AP_InertialSensor : AP_AccelCal_Client
+    float get_gyro_harmonic_notch_reference_freq_hz(void) const { return _harmonic_notch_filter.center_freq_hz(); }
+
+    // harmonic notch reference scale factor
+    float get_gyro_harmonic_notch_reference(void) const { return _harmonic_notch_filter.reference(); }
```
Are the 3 methods above going to be used soon? They are unused at the moment (as is the reference scale factor).

andyp1per replied:
Yes all used when we wire to throttle

rmackay9 replied:
It's a bit of a golden rule in AP that we don't add things before they are used so let's remove these.  Otherwise you can just look forward to a PR from muramura in 2 weeks removing them :-).

OXINARF commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -498,14 +505,27 @@ void AP_InertialSensor_Backend::update_gyro(uint8_t instance)
         _last_notch_bandwidth_hz[instance] = _gyro_notch_bandwidth_hz();
         _last_notch_attenuation_dB[instance] = _gyro_notch_attenuation_dB();
     }
+    // possily update the harmonic notch filter parameters
```
To follow the code in the _notify method, I'd prefer to see the code below be before the existing notch code.

andyp1per replied:
ok

OXINARF commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -202,6 +202,9 @@ class AP_InertialSensor : AP_AccelCal_Client
     uint8_t get_primary_gyro(void) const { return _primary_gyro; }
 
+    // Update the harmonic notch frequency
+    void update_harmonic_notch_freq_hz(float scaled_freq);
```
This isn't called anywhere at the moment, but I gather it's going to be used soon?

andyp1per replied:
Indeed. Already flying well with it.

OXINARF commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -498,14 +505,27 @@ void AP_InertialSensor_Backend::update_gyro(uint8_t instance)
+        _last_harmonic_notch_center_freq_hz[instance] = _gyro_harmonic_notch_center_freq_hz();
+    } 
+    // possily update the harmonic notch filter parameters
+    if (!is_equal(_last_harmonic_notch_bandwidth_hz[instance], _gyro_harmonic_notch_bandwidth_hz()) ||
```
This *if* should be before the one above, with the one above being an else to this one. Reason is that if entering this *if* then center frequency will already be reset (and if not, then check if it's needed to update the center frequency).

andyp1per replied:
Ok, I just copied the previous code, will fix :)

OXINARF commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -75,6 +75,9 @@ class AP_InertialSensor_Backend
     void notify_fifo_reset(void);
+
+    // update the harmonic notch frequency for the selected backend
+    void _update_harmonic_notch_freq_hz(uint8_t instance, float scaled_freq);
```
Not defined, should likely be removed.

andyp1per replied:
Ok

OXINARF commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -252,10 +266,15 @@ class AP_InertialSensor_Backend
+    // support for updating dynamic filter at runtime
+    float _last_harmonic_notch_center_freq_hz[INS_MAX_INSTANCES];
+    float _last_harmonic_notch_bandwidth_hz[INS_MAX_INSTANCES];
+    float _last_harmonic_notch_attenuation_dB[INS_MAX_INSTANCES];
```
This is wrong (as is for all the above). This is in the Backend class so it already is per-instance, no array needed. You should fix the code adding this and if you can make a **separate commit** fixing the other ones I'd be grateful!

andyp1per replied:
So I agree that this is the backend, but every single "last_" member variable follows this pattern (not written by me). So probably all of them should be changed or none of them - I just copied what was there already. What would you like me to do? Oh I think I see - you want all of them fixed but the ones not added by me in a separate commit?

OXINARF replied:
> Oh I think I see - you want all of them fixed but the ones not added by me in a separate commit?

Yes, that's it. Separate commit is for easily noting there was a previous bug, not introduced by you. As said, you doing commit is optional as the code is already there and you aren't forced to fix it - you are forced to fix your new code though 🙂

OXINARF commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -1737,6 +1746,14 @@ void AP_InertialSensor::acal_update()
+// Update the harmonic notch frequency
+void AP_InertialSensor::update_harmonic_notch_freq_hz(float scaled_freq) {
+    // When disarmed, throttle is zero
+    if (scaled_freq > 0) {
```
```suggestion
if (is_positive(scale_factor)) {
```

andyp1per replied:
Ok!

andyp1per commented:
> @andyp1per sorry for the delay, I really had quite a few outdated comments, it took a bit to unscramble my brain 😄

No problem. Sorry for asking you to look at so many commits!
 
> Besides the inline comments, I'm not very happy with the for loops for initializing and, especially, updating the filters array. Is the `update` method expected to be called a lot? I think we could 

As per Leonard we would update the filter parameters at 50Hz in the throttle_loop, so substantially below the rate at which the filter is applied. This works well and I have been flying with this without issue.

> probably use `__builtin_ffs` to make the for loop only go through the set bits (we need to check it works on Apple's Clang), but I'm not sure if that would be more efficient given that (for now) it's just 8 loops. Anyway, that's more of an optimization problem and can be done later if needed.

Yes, I think perhaps running 8 filters is quite ambitious, maybe there should be a lower limit! @lthall might have an opinion here. I have been running with one harmonic (two filters) without issue.

andyp1per commented on libraries/Filter/NotchFilter.h:
```diff
@@ -29,12 +29,17 @@
 template <class T>
 class NotchFilter {
 public:
+    template <class U> friend class HarmonicNotchFilter;
```
Yes. I went back on forth on making HarmonicNotch inherit from NotchFilter, but there is no need other than to make it a friend, also even with inheritance you still need a friend so that the HarmonicNotch can initialize the filter bank.

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -453,6 +453,10 @@ const AP_Param::GroupInfo AP_InertialSensor::var_info[] = {
 
+    // @Group: DNTCH_
+    // @Path: ../Filter/HarmonicNotchFilter.cpp
+    AP_SUBGROUPINFO(_harmonic_notch_filter, "DNTCH_",  41, AP_InertialSensor, HarmonicNotchFilterParams),
```
Maybe - ultimately we want to wire this up to the throttle output so that the center frequency varies dynamically. Happy to go with whatever the consensus is here. Originally the class was called DynamicNotchFilter.

OXINARF replied:
> Originally the class was called DynamicNotchFilter.

Yeah, that's why I thought this could be changed too. I don't really have a strong opinion on either. @lthall?

lthall replied:
I would say go with HNTCH

andyp1per replied:
Fixed

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -471,9 +475,14 @@ AP_InertialSensor::AP_InertialSensor() :
     _singleton = this;
     AP_Param::setup_object_defaults(this, var_info);
+
+    // This is not user configurable, but start with the configured default.
```
Ok, I'll improve the comment!

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -471,9 +475,14 @@ AP_InertialSensor::AP_InertialSensor() :
     for (uint8_t i=0; i<INS_MAX_INSTANCES; i++) {
         _gyro_cal_ok[i] = true;
         _accel_max_abs_offsets[i] = 3.5f;
+        _gyro_harmonic_notch_filter[i].create(_harmonic_notch_filter.harmonics());
```
Ouch, ok!

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -24,7 +24,7 @@
 #include <Filter/LowPassFilter2p.h>
 #include <Filter/LowPassFilter.h>
-#include <Filter/NotchFilter.h>
+#include <Filter/HarmonicNotchFilter.h>
```
Ok!

OXINARF commented:
@andyp1per 

> As per Leonard we would update the filter parameters at 50Hz in the throttle_loop, so substantially below the rate at which the filter is applied. This works well and I have been flying with this without issue.

Ok, I think it's not too bad. As I said, this was more of a comment than something you need to change.

> Yes, I think perhaps running 8 filters is quite ambitious, maybe there should be a lower limit! @lthall might have an opinion here. I have been running with one harmonic (two filters) without issue.

Yeah, 8 is likely ambitious but I wouldn't stress too much about limiting it.

Do you have the below on your browser?

![image](https://user-images.githubusercontent.com/1778527/61976252-30a3f800-afe3-11e9-80a4-e6780ac344c6.png)

You seem to have passed over those 18 comments (don't worry, I've seen this happen many times).

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -243,6 +243,17 @@ class AP_InertialSensor_Backend
 
     uint8_t _gyro_notch_enabled(void) const { return _imu._notch_filter.enabled(); }
 
+    // return the dynamic notch filter center in Hz for the sample rate
```
Ok

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -243,6 +243,17 @@ class AP_InertialSensor_Backend
+    // return the notch filter bandwidth in Hz for the sample rate
+    uint16_t _gyro_harmonic_notch_bandwidth_hz(void) const { return _imu._harmonic_notch_filter.bandwidth_hz(); }
+
+    // return the notch filter attenuation in dB for the sample rate
```
ok

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -256,6 +267,11 @@ class AP_InertialSensor_Backend
     uint16_t _last_notch_bandwidth_hz[INS_MAX_INSTANCES];
     float _last_notch_attenuation_dB[INS_MAX_INSTANCES];
 
+    // support for updating dynamic filter at runtime
```
ok

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -65,73 +65,83 @@ const AP_Param::GroupInfo HarmonicNotchFilterParams::var_info[] = {
-    filters = new NotchFilter<T>[_harmonics+1];
-    harmonics = _harmonics;
+    _num_filters = 1;
+    for (uint8_t i = 0; i < HNF_MAX_FILTERS-1; i++) {
```
Because you always get the first harmonic if its enabled, the parameter controls the additional harmonics, but the maximum includes all

andyp1per replied:
I take that back, you are right - I have fixed now

andyp1per commented on libraries/Filter/HarmonicNotchFilter.h:
```diff
@@ -20,6 +20,7 @@
 #include <AP_Param/AP_Param.h>
 #include "NotchFilter.h"
 
+#define HNF_MAX_FILTERS 8
```
ok

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -65,73 +65,83 @@ const AP_Param::GroupInfo HarmonicNotchFilterParams::var_info[] = {
-        filters[i].internal_init(sample_freq_hz, center_freq_hz * (i+1), A, Q);
+    for (uint8_t i=0; i<HNF_MAX_FILTERS; i++) {
+        if ((1U<<i) & _harmonics) {
+            _filters[i].internal_init(sample_freq_hz, center_freq_hz * (i+1), _A, _Q);
```
Good catch

andyp1per commented on libraries/Filter/NotchFilter.h:
```diff
@@ -59,15 +59,15 @@ class NotchFilterParams {
-    AP_Int16 _center_freq_hz;
-    AP_Int16 _bandwidth_hz;
+    AP_Float _center_freq_hz;
+    AP_Float _bandwidth_hz;
```
@lthall asked me to do this based on a conversation with @tridge - what is the right way to manage this change?

lthall replied:
Ahh, this is where we move the parameter to the last position and give it a new index.

lthall replied:
Unfortunately I don't know how to do this with those numbers. @OXINARF, is there any chance you could push a fix for this one?

andyp1per replied:
I have changed the indexes but don't know how to do the conversion code

rmackay9 replied:
we've added the conversion code now so closing this.

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,169 @@
+    // @User: Advanced
+    AP_GROUPINFO("ATT", 4, HarmonicNotchFilterParams, _attenuation_dB, 15),
+
+    // @Param: HMNCS
```
Instead of HMNCS? I guess I find that confusing since people might think they actually should enter a frequency here

lthall replied:
I think HMNCS is more appropriate here.

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,169 @@
+
+    T output = _filters[0].apply(sample);
+    for (uint8_t i=1; i<_num_filters; i++) {
+        output = _filters[i].apply(output);
```
ok

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,169 @@
+        return;
+    }
+
+    for (uint8_t i=0; i<_num_filters+1; i++) {
```
Good catch

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,169 @@
+/* 
+   instantiate template classes
+ */
+template class HarmonicNotchFilter<float>;
```
ok

andyp1per commented on libraries/Filter/HarmonicNotchFilter.h:
```diff
@@ -0,0 +1,63 @@
+    AP_Float _reference;
+};
+
+typedef HarmonicNotchFilter<float> HarmonicNotchFilterFloat;
```
ok

andyp1per commented on libraries/Filter/NotchFilter.h:
```diff
@@ -29,12 +29,23 @@
+    }
+
 private:
+    void internal_init(float sample_freq_hz, float center_freq_hz, float A, float Q);
```
ok

andyp1per commented on libraries/Filter/NotchFilter.h:
```diff
@@ -29,12 +29,23 @@
     void reset();
 
+    // calculate attenuation and quality from provided center frequency and bandwidth
+    static void calculate_A_and_Q(float center_freq_hz, float bandwidth_hz, float attenuation_dB, float& A, float& Q) {
```
wanted to inline the definition ...

andyp1per commented on libraries/Filter/NotchFilter.h:
```diff
@@ -29,12 +29,23 @@
+    static void calculate_A_and_Q(float center_freq_hz, float bandwidth_hz, float attenuation_dB, float& A, float& Q) {
+        const float octaves = log2f(center_freq_hz / (center_freq_hz - bandwidth_hz / 2.0f)) * 2.0f;
+        A = powf(10, -attenuation_dB / 40.0f);
+        Q = sqrtf(powf(2, octaves)) / (powf(2,octaves) - 1.0f);
```
ok

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -211,6 +214,15 @@ class AP_InertialSensor : AP_AccelCal_Client
+    float get_gyro_harmonic_notch_center_freq_hz(void) const { return _calculated_harmonic_notch_freq_hz; }
+
+    // harmonic notch reference center frequency
+    float get_gyro_harmonic_notch_reference_freq_hz(void) const { return _harmonic_notch_filter.center_freq_hz(); }
```
Hmmn, ok. Will have to think about this.

andyp1per replied:
The harmonic notch itself doesn't really know anything about the reference vs center frequency - it only has one. It is the IMU code that is making a distinction. So I'm comfortable with the naming but could change them to center_freq_hz and current_center_freq_hz if you think that would be better?

lthall replied:
Maybe this should be:

    float get_gyro_dynamic_notch_center_freq_hz(void) const { return _calculated_harmonic_notch_freq_hz; }

     // harmonic notch reference center frequency
    float get_gyro_harmonic_notch_center_freq_hz(void) const { return _harmonic_notch_filter.center_freq_hz(); }

We only need the dynamic call because we are using it dynamically otherwise we would just use the second.

andyp1per replied:
Ok done

andyp1per commented:
@OXINARF ok I have fixed the ones I can - please mark my homework :) I am on vacation now for a week so won't be able to fly with the changes just yet (or make further changes). If you can @lthall can figure out the naming and parameter transition that would be great!

lthall commented:
Yes, I suspect we will only ever need 3 of these filters to take care of the prop noise frequencies. The flexibility is very useful though as depending on the structure of the aircraft we may have different harmonics present. In some cases we don't actually need the first harmonic at all! In other cases I could see us needing the 1st, 3rd, and 6th.

andyp1per commented:
> Yes, I suspect we will only ever need 3 of these filters to take care of the prop noise frequencies. The flexibility is very useful though as depending on the structure of the aircraft we may have different harmonics present. In some cases we don't actually need the first harmonic at all! In other cases I could see us needing the 1st, 3rd, and 6th.

Ok, I have changed it to only allow 3 filters total but their location can still be selected via the bitmask.

andyp1per commented:
@OXINARF I have addressed all of your comments apart from the conversion code for the variable type change - let me know how to do that or how you want to manage that. Ready for @tridge I think maybe!

guglie commented:
> we will only ever need 3 of these filters to take care of the prop noise frequencies

Hi @lthall, why is that? Is it related to the quadcopter motor control?
Can this be the reason why I can see only 3 prop frequencies in the [gyro spectrum](https://discuss.ardupilot.org/t/imu-filter-tool/43633) of a flight?
(the 3 green/yellow lines close to each other, with their 3 harmonics above)

<img width="821" alt="gyroscope X spectrum" src="https://user-images.githubusercontent.com/438683/62037727-a7640f80-b1f4-11e9-8a9f-377acb5cc2ed.png">


By the way thanks @andyp1per and @lthall for this great work!

lthall commented:
> Hi @lthall, why is that? Is it related to the quadcopter motor control?

There are three main signals from propeller interaction.
The axial unbalance of the props (rpm)
The blade passage frequency or the impact of air between each blade and the aircraft arm (rpm x no_blades)
The interaction between the upper and lower propeller blades. (rpm x no_blades^2)

After that we may see a second harmonic of one of these sources as well but generally by the time we have got any higher than this we are being attenuated by the 2p filter and don't have to worry about notching it out.

So 3 is not a hard and fast rule but in my experience 2 notch filters will do a great job with very occasionally 3 needed. However, it is possible that on some aircraft the may be some benefit to using additional harmonics.

OXINARF commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -88,8 +89,8 @@ void HarmonicNotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, fl
-    for (uint8_t i=0, filt=0; i<HNF_MAX_FILTERS; i++) {
-        if ((1U<<i) & _harmonics) {
+    for (uint8_t i = 0, filt = 0; i < HNF_MAX_HARMONICS && filt < _num_filters; i++) {
+        if (i == 0 || (1U<<i) & _harmonics) {
```
```suggestion
        if (i == 0 || (1U<<(i - 1)) & _harmonics) {
```

andyp1per replied:
I don't agree this correct unless we allow the bit mask to be for all harmonics rather than the ones above the fundamental

OXINARF replied:
The bitmask starts at bit 0 with the second harmonic. With current code you are making bit 0 not be used and bit 1 be 2nd harmonic (instead of 3rd), etc.

lthall replied:
Nice catch. I did want the first harmonic to be turned off. For example when the props are well balanced we may not need to turn on the first harmonic.

andyp1per replied:
So I have changed to allow control all harmonics. I believe your comment above to be incorrect - i starts at 0, so 1<<(i-1) is 1<<-1 which is invalid. You want to start with 1<<0 which is the first bit.

OXINARF replied:
@andyp1per You are missing that when i is 0 then the first condition was used. Loop was this:

- i = 0; i == 0 => true; second condition isn't run
- i = 1; i == 0 => false; (1<<1) & _harmonics
- i = 2; i == 0 => false; (1<<2) & _harmonics
- etc.

So no check of bit 0 in _harmonics.

Anyway, that's now changed with control for all harmonics.

OXINARF commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -122,8 +123,8 @@ void HarmonicNotchFilter<T>::update(float center_freq_hz)
-    for (uint8_t i=0, filt=0; i<HNF_MAX_FILTERS; i++) {
-        if (i==0 || (1U<<i) & _harmonics) {
+    for (uint8_t i = 0, filt = 0; i < HNF_MAX_HARMONICS && filt < _num_filters; i++) {
+        if (i == 0 || (1U<<i) & _harmonics) {
```
Same.

andyp1per replied:
Ditto

OXINARF commented on libraries/Filter/NotchFilter.cpp:
```diff
@@ -90,30 +90,30 @@ const AP_Param::GroupInfo NotchFilterParams::var_info[] = {
     // @Units: Hz
     // @User: Advanced
-    AP_GROUPINFO("BW", 3, NotchFilterParams, _bandwidth_hz, 20),
+    AP_GROUPINFO("BW", 5, NotchFilterParams, _bandwidth_hz, 20),
```
5 -> 6

andyp1per replied:
Fixed

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -103,7 +104,7 @@ template <class T>
 {
     _num_filters = 1;
-    for (uint8_t i = 0; i < HNF_MAX_FILTERS-1; i++) {
+    for (uint8_t i = 0; i < HNF_MAX_HARMONICS && _num_filters < HNF_MAX_FILTERS; i++) {
```
Actually since we are now limiting the number of filters I can just change the docs to allow the 8th bit. Also pretty sure I get 3 filters above - yes I only get 2 extra harmonics but with the fundamental it's 3 total.

OXINARF replied:
> Actually since we are now limiting the number of filters I can just change the docs to allow the 8th bit.

Just changing the docs doesn't work though because other places only allow 8 (including the first harmonic).

> Also pretty sure I get 3 filters above - yes I only get 2 extra harmonics but with the fundamental it's 3 total.

You are right!

andyp1per replied:
Ok as per @lthall all harmonics can now be controlled. I have made the default the first two.

andyp1per commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -53,7 +54,7 @@ const AP_Param::GroupInfo HarmonicNotchFilterParams::var_info[] = {
     // @DisplayName: Harmonics
-    // @Description: Bitmask of harmonic frequencies to add to the filter fundamental frequency. This option takes effect on the next reboot.
+    // @Description: Bitmask of harmonic frequencies to add to the filter fundamental frequency. This option takes effect on the next reboot. A maximum of 3 harmonics can be used at any one time
     // @Bitmask: 0:2nd harmonic,1:3rd harmonic,2:4th harmonic,3:5th hamronic,0:6th harmonic,0:7th harmonic,0:8th harmonic
```
Fixed

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -455,7 +455,7 @@ const AP_Param::GroupInfo AP_InertialSensor::var_info[] = {
 
     // @Group: DNTCH_
```
Fixed

lthall commented:
![image](https://user-images.githubusercontent.com/100896/62091670-42222400-b2b1-11e9-8157-0d69375551cb.png)

Just to give people an idea of what we see on a heli.

andyp1per commented:
Ok I will fix

Sent from my iPhone

> On 30 Jul 2019, at 00:23, Leonard Hall <notifications@github.com> wrote:
> 
> @lthall commented on this pull request.
> 
> In libraries/Filter/HarmonicNotchFilter.cpp:
> 
> > @@ -88,8 +89,8 @@ void HarmonicNotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, fl
>  
>      NotchFilter<T>::calculate_A_and_Q(center_freq_hz, bandwidth_hz, attenuation_dB, _A, _Q);
>  
> -    for (uint8_t i=0, filt=0; i<HNF_MAX_FILTERS; i++) {
> -        if ((1U<<i) & _harmonics) {
> +    for (uint8_t i = 0, filt = 0; i < HNF_MAX_HARMONICS && filt < _num_filters; i++) {
> +        if (i == 0 || (1U<<i) & _harmonics) {
> Nice catch. I did want the first harmonic to be turned off. For example when the props are well balanced we may not need to turn on the first harmonic.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

andyp1per commented:
@OXINARF all done I believe

andyp1per commented:
As long as you are ok with how it is now 😀

Sent from my iPhone

> On 30 Jul 2019, at 10:35, Francisco Ferreira <notifications@github.com> wrote:
> 
> @OXINARF commented on this pull request.
> 
> In libraries/Filter/HarmonicNotchFilter.cpp:
> 
> > @@ -88,8 +89,8 @@ void HarmonicNotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, fl
>  
>      NotchFilter<T>::calculate_A_and_Q(center_freq_hz, bandwidth_hz, attenuation_dB, _A, _Q);
>  
> -    for (uint8_t i=0, filt=0; i<HNF_MAX_FILTERS; i++) {
> -        if ((1U<<i) & _harmonics) {
> +    for (uint8_t i = 0, filt = 0; i < HNF_MAX_HARMONICS && filt < _num_filters; i++) {
> +        if (i == 0 || (1U<<i) & _harmonics) {
> @andyp1per You are missing that when i is 0 then the first condition was used. Loop was this:
> 
> i = 0; i == 0 => true; second condition isn't run
> i = 1; i == 0 => false; (1<<1) & _harmonics
> i = 2; i == 0 => false; (1<<2) & _harmonics
> etc.
> So no check of bit 0 in _harmonics.
> 
> Anyway, that's now changed with control for all harmonics.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

OXINARF commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,170 @@
+            _num_filters++;
+        }
+    }
+    _filters = new NotchFilter<T>[_num_filters];
```
```suggestion
    if (_num_filters > 0) {
        _filters = new NotchFilter<T>[_num_filters];
    }
```
There is no reason to allocate (and initialize) if there is no harmonic set in the bitmask.

By the way, we need to add a destrutor to delete the array, otherwise we have memory leak.

andyp1per replied:
Ok, I found a pub with power, so laptop charging nicely, beer is cold - and I can push this.

OXINARF commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -108,13 +108,17 @@ void HarmonicNotchFilter<T>::create(uint8_t harmonics)
 template <class T>
 HarmonicNotchFilter<T>::~HarmonicNotchFilter() {
-    delete[] _filters;
+    if (_num_filters > 0) {
```
This check isn't needed, it's safe to delete a nullptr variable.

andyp1per replied:
True, but unless you feel very strongly about this I prefer the explicitness.

rmackay9 replied:
I agree with oxinarf actually that just do the delete and maybe also set the _num_filters = 0 if that makes sense.

rmackay9 commented:
Sorry for the delay but I've added a commit in this branch which does the parameter conversion for INS_NOTCH_FREQ and INS_NOTCH_BW. https://github.com/rmackay9/rmackay9-ardupilot/commits/andypiper-harmonic-filter

For future reference the way I figured out the magic numbers for the param conversion is:

- set the AP_PARAM_KEY_DUMP definition to "1" [here in AP_Param.h](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Param/AP_Param.h#L36)
- change the delay here from 1ms to 2ms (not sure if this is required but I think it is) [here in AP_Param::show_all](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Param/AP_Param.cpp#L2338)
- removed the #if / #end from [here in Copter's system.cpp](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/system.cpp#L262) (the definitions don't seem to work)
- started the old code in SITL and all the parameter names and their magic numbers are displayed
- did some copy-pasting from Copter's Parameters.cpp file's [existing parameter conversion tables like the ones done for the attitude controller's filters](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/Parameters.cpp#L1219).

andyp1per commented:
Thanks @rmackay9 ! Should I pull these changes into my branch?

andyp1per commented:
@rmackay9 I've rebased on master and mirrored your changes here

andyp1per commented:
@OXINARF are you able to approve these changes now?

rmackay9 commented:
Great.  I'd like ot go through this over the next few days and just make sure I can't find any issues.. Sorry for the delay..

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,179 @@
+const AP_Param::GroupInfo HarmonicNotchFilterParams::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Enable
```
The DisplayName should probably be extended to be something more specific so that it is clear to the user what it is for.  So for example "Harmonic Filter Enable".  The description could be the same.  In general I think it is a good idea to put the feature or subsystem name at the beginning.

andyp1per replied:
Fixed

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,179 @@
+    AP_GROUPINFO_FLAGS("ENABLE", 1, HarmonicNotchFilterParams, _enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: FREQ
+    // @DisplayName: Base Frequency
```
This should also have "Harmonic Filter" or something similar appended to the end.  Basically all the parameters should have this so let's add it to all of them.

andyp1per replied:
Fixed

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,179 @@
+
+
+/*
+  initialise filter
```
It would be nice to have comments above each function describing what it's for.  At least one line.  In the case of "init" and "create" it looks like it's copy-pasted from somewhere else.  Better comments help the next developer (and even the original developer) when they have to figure out how something is working.

andyp1per replied:
changed to allocate_filters and added comments

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,179 @@
+    // adjust the center frequency to be in the allowable range
+    center_freq_hz = constrain_float(center_freq_hz, bandwidth_hz * 0.52f, sample_freq_hz * 0.48f);
+
+    NotchFilter<T>::calculate_A_and_Q(center_freq_hz, bandwidth_hz, attenuation_dB, _A, _Q);
```
A little more commenting inside the functions would also help save developer time in the future.

andyp1per replied:
done

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,179 @@
+}
+
+template <class T>
+HarmonicNotchFilter<T>::~HarmonicNotchFilter() {
```
sorry to be pedantic but destructors are normally at the top in both the .h and .cpp files.

andyp1per replied:
done

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,179 @@
+  initialise filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::create(uint8_t harmonics)
```
"create" is not a word that we commonly use.  Normally we have a constructor (method named the same as the class) and an "init" but I've never seen a "create" before.  For consistency and ease-of-understanding for new developers I think it's a good idea to avoid new terms and words if possible.

Another small pedantic point is I'd prefer if we can make the order of the methods the same in the .h and .cpp file.  So for example if "init" is declared before "create" in the .h then it should be in the same order in the .cpp file.

andyp1per replied:
changed to allocate_filters and moved things around as requested

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.h:
```diff
@@ -0,0 +1,61 @@
+private:
+    NotchFilter<T>*  _filters;
+    float _sample_freq_hz;
+    float _A;
```
It would be very helpful for future developers to have comments beside each member declaration giving some idea of what these are for.

andyp1per replied:
done

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.h:
```diff
@@ -0,0 +1,61 @@
+class HarmonicNotchFilterParams : public NotchFilterParams {
+public:
+    HarmonicNotchFilterParams(void);
+    void set_center_freq_hz(float center_freq) { _center_freq_hz.set(center_freq); }
```
a single line comment above each method would be good I think.

andyp1per replied:
done

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.h:
```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <AP_Math/AP_Math.h>
+#include <cmath>
```
I don't think we need the include of <cmath> and <inttypes.h> do we?

andyp1per replied:
pasted from NotchFilter, removed now

rmackay9 commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -1751,6 +1765,14 @@ void AP_InertialSensor::acal_update()
 
+// Update the harmonic notch frequency
+void AP_InertialSensor::update_harmonic_notch_freq_hz(float scaled_freq) {
+    // When disarmed, throttle is zero
```
this is a bit of a funny comment here I think.. copy-paste issue?
Also I don't think that this method, update_harmonic_notch_freq_hz, is ever called so I think it would be best to remove it for now.

andyp1per replied:
So the issue here is that we want to add the harmonic update PR immediately after this and if we remove it here it means AP_InertialSensor is in that PR and review. I guess the thinking from @lthall and I was get all the difficult reviews out the way here and then the follow-on PR becomes much simpler. Or we add that code in here. Or we leave it as is. So your call from a process perspective. Also the update code is more naturally associated with the harmonic filter code.

rmackay9 replied:
if it's coming in the next few weeks then OK.. if it's longer than that then let's put it in the next PR..

andyp1per replied:
Cool - it's ready to go. I have been flying with it for several weeks now. So I'll leave this in.

andyp1per replied:
Changed the comment

rmackay9 commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -493,18 +500,31 @@ void AP_InertialSensor_Backend::update_gyro(uint8_t instance)
+        _last_harmonic_notch_center_freq_hz = _gyro_harmonic_notch_center_freq_hz();
+        _last_harmonic_notch_bandwidth_hz = _gyro_harmonic_notch_bandwidth_hz();
+        _last_harmonic_notch_attenuation_dB = _gyro_harmonic_notch_attenuation_dB();
+    }
```
small formatting issue here. the "}" should be on the same line as the "else"

andyp1per replied:
Fixed

rmackay9 commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -493,18 +500,31 @@ void AP_InertialSensor_Backend::update_gyro(uint8_t instance)
+    }
+
+    // possily update the harmonic notch filter parameters
+    if (!is_equal(_last_harmonic_notch_bandwidth_hz, _gyro_harmonic_notch_bandwidth_hz()) ||
```
it's not super important but if it were me I would probably move this into a method in the harmonic filter itself.  A self check of whether it's parameter values have changed.  this isn't a blocker though..

andyp1per replied:
That's quite invasive since I have to move all the _last variables there as well. If its ok I'll leave as is.

rmackay9 commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -233,29 +233,45 @@ class AP_InertialSensor_Backend
+    float _gyro_harmonic_notch_center_freq_hz() const { return _imu._calculated_harmonic_notch_freq_hz; }
+
+    // return the harmonic notch filter bandwidth in Hz for the sample rate
+    float _gyro_harmonic_notch_bandwidth_hz(void) const { return _imu._harmonic_notch_filter.bandwidth_hz(); }
```
sorry to keep moaning but I'm really not a fan of underscores in the front of method names.  For members/variables it's ok but not for the methods.  I see you're just following the pattern here so it's not a blocker.

andyp1per replied:
Removed.

andyp1per commented:
> In general it looks pretty good. I've added some requests for a number of non-functional things.
> 
> Sorry for the delay in reviewing

@rmackay9 thanks for the review! I don't mind the pickiness - it helps me learn house style for next time ;)

tridge commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,193 @@
+        }
+    }
+    if (_num_filters > 0) {
+        _filters = new NotchFilter<T>[_num_filters];
```
what happens if this allocation fails? We should at least show an error msg. Are all code paths that may use _filters ok with nullptr?

andyp1per replied:
Good catch. I will fix. Yes its safe because we never set _initialised if there is a nullptr.

tridge commented on libraries/Filter/NotchFilter.cpp:
```diff
@@ -72,30 +90,30 @@ const AP_Param::GroupInfo NotchFilterParams::var_info[] = {
     // @Range: 10 400
     // @Units: Hz
     // @User: Advanced
-    AP_GROUPINFO("FREQ", 2, NotchFilterParams, _center_freq_hz, 80),
```
a comment on why 2 and 3 are unused (and reserved) would be good

andyp1per replied:
Done

tridge commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -0,0 +1,193 @@
+template <class T>
+void HarmonicNotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, float bandwidth_hz, float attenuation_dB)
+{
+    if (_filters == nullptr) {
```
as far as I can tell this is only called once, and it is called before _filters is allocated, so this function never sets _initialised to true, so the filter is never activated.

andyp1per replied:
It's called repeatedly in AP_InertialSensor_Backend::update_gyro() if the settings change. allocate_filters() is called in AP_InertialSensor::init() which is called before the scheduler has started calling update (even though the backends are allocated earlier). That said in my testing I am updating the center_freq repeatedly, so if there was an issue I wouldn't see it - let me double check in SITL.

andyp1per replied:
You are completely correct - fixed now!

tridge replied:
thanks, I do wonder how much of the testing this invalidates though. The filter didn't do anything when I tested it

andyp1per replied:
the irony is that it was one of the many reviews that introduced the problem. So far the main source of in-flight testing has been me. So I have done some basic aggressive flying testing with the fixes below but not acro-based testing that I was doing before - I will do this again when the weather improves. For CPU testing, what is the procedure there?

andyp1per replied:
Are averaged timestamps ok? I've done that successfully for my FFT change. Do you have a ballpark latency in mind? I have F4 (omnibusf4pro, pixracer and auavx2) and F7 (kakute) boards to test on.

andyp1per commented:
@tridge @rmackay9 if the changes are good I will squash once more

andyp1per commented:
Don't merge this just yet. I need to do some additional testing - I was getting some funny behavior today where the gyro had no data in stabilize - and my copter flipped and crashed - but was fine in AltHold and PosHold, but I am flying with a lot of custom changes, so it may simply be something else. I'm just suspicious of this initialization change.

andyp1per commented:
It seems to me that _gyro_raw_sample_rate can change over time, but we do not deal with that in any of the filters sampling frequency. Is the assumption that it won't change *much*? I'm just wondering whether there is any kind of big change at startup with some sensors - my change is the first that assumes that this is set right at initialization.

andyp1per commented:
Found it! I at first thought it was https://github.com/ArduPilot/ardupilot/pull/12183 but had to create https://github.com/ArduPilot/ardupilot/pull/12182 to allow me to see if I was hitting gyro errors or not (I wasn't). I also discovered a bug in the process whereby if we detect NaN or Inf in the filtered gyro values we reset the filters but still accept the invalid output, but that wasn't the issue. The main issue was that we were allowing fundamental center frequencies that caused the center frequencies of the *harmonics* to go above the Nyquist frequency. If that happened then all hell broke lose. It was quite subtle and difficult to reproduce but got it eventually. The incoming fix fixes both this, the reset bug above and the issue with sensor rates changing https://github.com/ArduPilot/ardupilot/pull/11795#issuecomment-526485212 
@rmackay9 and @tridge the PR is ok to merge now

andyp1per commented:
Weather forecast for tomorrow is ok. I'll do some blasting.

andyp1per commented:
@tridge incidentally on the CPU cost the critical path - apply() - has been designed to be a simple loop through the notch filter apply()'s so the cost should simply be the cost of a notch x the number of notches. I guess we could try unrolling the loop, but might not achieve much. It would, however, be interesting to know in general how much the notches cost.

andyp1per commented:
Well good job you asked @tridge, because the code I submitted was after I had tested in the field, then reviewed, then made a small change to fix a problem I could see - which then re-broke the nyquist stuff. I have submitted the fix and now it works great. Here is an FFT graph from flying around aggressively in Acro:

![image](https://user-images.githubusercontent.com/2893260/64079409-198daf80-ccdf-11e9-9bcd-35404053d00b.png)

You can see that the gyro noise has been completely flattened across the spectrum, and interestingly there is a knock-on effect with the accels. I made the fix and flew 10 batteries in quite windy conditions - all was great, so I'm very confident now in this fix.

bnsgeyer commented:
@andyp1per  was looking to try out this feature but have a question about the parameter REF.  What exactly is it used for?  The description did not help me in determining how to set this parameter. 
Thanks,
Bill

andyp1per commented:
@bnsgeyer I don't think there is much point in trying it until the dynamic center frequency control is added. REF determines the scaling of this, but without the follow-on PR does not currently do anything. Adding the dynamic center frequency control is a relatively small change which I plan to open as soon as this change in.
If you are feeling lucky you can try my 3.6 version that rolls all these together (including FFT) - https://github.com/andyp1per/ardupilot/tree/dynamic-filtering-3.6-extra-experimental - but the name highlights the caveat emptor and I would not try it on anything of value :)

bnsgeyer commented:
@andyp1per what do you mean that it doesn’t do anything?
Heli rotor speeds are very constant when using a governor. And even when we don’t use a governor users set the throttle with collective to keep rotor speed constant. The only reason we might need the dynamic centering is if we get into an autorotation where rotor speed is not that constant.  Then the dynamic FFT would be useful. So I don’t have to concern myself with the reference parameter. I just want it to work for the freq I set and harmonics of that.

andyp1per commented:
That's fine then - you don't need to worry about REF - it's only for dynamic centering - just set INS_HNTCH_FREQ  and friends appropriately

andyp1per commented:
rebased the nyquist changes in order to kick travis

andyp1per commented:
@tridge I tested the performance - the maths is pretty easy it basically costs 2us per notch on an F4. So without any notches the total gyro/accel cycle time is ~135us (this was on a MPU6000) and the gyro filter part of that is ~6us. If you add the static notch that goes up to ~8us. If you then add the harmonic notch first harmonic that goes up to ~10us, 2nd harmonic ~12us, 3rd harmonic - ~14us which is the maximum number we allow. I'll cook up a PR for what I used to measure this as I think it is generally useful. If you want me to test on a F7 I can or any other combinations.

andyp1per commented:
I used this https://github.com/ArduPilot/ardupilot/pull/12191 to measure the timing

bnsgeyer commented:
@andyp1per. Just tested the harmonic notch. Could not tell too much of a difference in the aircraft flight but just did a short hop. But looked at the FFT for post filter and WOW! What a difference. I will post more later.  And I want to make sure I set up the batch log correctly for post filter sampling.  I am pretty sure I got it.  
Great job to you and @lthall!!!! Great contribution.

bnsgeyer commented:
So initially I set the harmonic filter to attenuate the 2/rev, 4/rev, and 10/rev as I had seen in the raw batch log data.  Up to this point I had never collected the post filtered data using the batch log.  Probably because this only became available in master and I hadn't flown it much.  Here are the settings for the HNTCH.  FREQ=50; BW=10; ATT=15; HMNCS=11  Here is the FFT
![image](https://user-images.githubusercontent.com/9907906/64138908-0e18b200-cdcd-11e9-89c6-ee02493ebfac.png)
I guess I was wondering why the 50 hz (2/rev) wasn't attenuated in this case.  So I did two more flights.  One with no HNTCH and one with HNTCH only attenuating the 2/rev and the 4/rev.  So here is the FFT with no HNTCH but using the NOTCH set for the 1/rev (25 hz).  
![image](https://user-images.githubusercontent.com/9907906/64139112-cb0b0e80-cdcd-11e9-80fa-90317a2a6b64.png)

And here is the FFT with HNTCH attenuating the 2/rev and the 4/rev. 
![image](https://user-images.githubusercontent.com/9907906/64139184-fdb50700-cdcd-11e9-95b7-fae5f76df034.png)
 It was definitely working here to suppress the 2/rev.

So I wonder, have you used it to suppress more than just the first 3 consecutive.  From what I've read in this PR, the user can specify what ever harmonics up to 3 different harmonics.  Is that correct?  Please let me know if I got the input correct for the HMNCS bitmask in my first flight (first plot above).

Now that I know the higher harmonics (8, 10, and 12/rev) are being attenuated well.  I think I will disable NOTCH and use the HNTCH to suppress the 1, 2 and 4/rev (25, 50, and 100 hz).

andyp1per commented:
The suppression is in integer multiples of the fundamental frequency that you specify, so for 50Hz and 11 as the harmonics that should be 50, 150, 200. Note that if you are running at the standard loop rate of 400Hz then 200 will be disabled because it is close to the nyquist frequency. That still doesn't explain the behaviour your observe in the first graph - I agree it should be attenuating 50Hz. Also the default notch settings are not very aggressive. On my racer I use an ATT of 100 and a BW of 65Hz for a 195Hz center.

lthall commented:
Hey Andy, aren't you getting mixed up with the frequencies there. 

Note that if you are running at the standard loop rate of 400Hz without fast sampling then 200 will be disabled because it is close to the nyquist frequency. 

The notch filter should be running at the gyro sample frequency so the nyquist will be half that or something like 500Hz.

andyp1per commented:
> The notch filter should be running at the gyro sample frequency so the nyquist will be half that or something like 500Hz.

Sorry, my mistake - you are right it is always a minimum of 1Khz, which is great!

andyp1per commented:
@bnsgeyer I have a tone generator and vibration speaker so can try and replicate your setup this evening. Just one question - you definitely rebooted when changing the harmonics?

bnsgeyer commented:
@andyp1per I think it was my mistake. I did not reboot after changing the parameters. So I am sure it was running with the default parameters for the first flight. I did not realize you needed to reboot after changing any of the parameters other than ENABLE.  Sorry for the false alarm. 
As far as the harmonics with a bitmask of 11, wouldn’t it suppress the 1, 2 and 4 harmonics (50, 100, and 200). Aren’t those the bits that make up 11.  You said it would suppress 1, 3. and 4 (50, 150, 200).   That would be a bitmask of 13, right?

andyp1per commented:
@bnsgeyer sorry, got my binary backwards - yes 50, 100, 200. The only parameter that needs a reboot is the harmonics as this affects the filter allocation -  all of the others are dynamic.

andyp1per commented:
Ok, so this is not strictly necessary, but I want to brag a little about my setup at this point as it's pretty cool and also instructive :)
My vibration test bed has evolved a little from @priseborough 's and I now have a OmnibusF4Pro and KakuteF7 hard mounted to a PDB which I have stuck to a vibration speaker being driven by a bluetooth amp:

![IMG_1657](https://user-images.githubusercontent.com/2893260/64204482-c60d9400-ce8d-11e9-9b02-ab06815aa003.jpg)

The omnibus is on top and I used that for these experiments as it has no sensor damping. I have been using this setup for my dynamic FFT development in https://github.com/ArduPilot/ardupilot/pull/11886 but that's a different story. 
I discovered (MultiTone Pro)[https://apps.apple.com/us/app/multitone-generator-pro/id1143928697] which allows me to drive multiple frequencies through the speaker. In this experiment I selected 50Hz, 100Hz, 150Hz and 200Hz as per @bnsgeyer 's setup and ran those through at the same volume and then took the post-filter batch output:

![image](https://user-images.githubusercontent.com/2893260/64203703-f94f2380-ce8b-11e9-9e3a-9c922fbb3f18.png)

You can see the tones clearly standing out and the roll-off is because of the gyro LPF. I then enabled the harmonic notch using @bnsgeyer 's settings to yield:

![image](https://user-images.githubusercontent.com/2893260/64204007-be012480-ce8c-11e9-9558-971555100a3b.png)

You can see that 50, 100 and 200 are being killed fairly effectively but 150 is preserved as expected. I then increase the ATT to 100dB to yield this:

![image](https://user-images.githubusercontent.com/2893260/64204212-2ea84100-ce8d-11e9-8b11-8b525043701e.png)

You can see that the peaks have now been completely eliminated. So the notch is working really well! Looking forward to the merge @tridge ;)

andyp1per commented:
I rechecked on my test bed with Leonard's filter tweaks and then did a short test flight pushing the throttle above the magic 250Hz briefly - all looks fine. @lthall good to go

rmackay9 commented:
I see the HarmonicNotchFilter is created as a template but is there any plan to make it work with anything besides a Vector3f?  This is not necessarily a blocker but just a question.

rmackay9 commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -225,22 +225,28 @@ void AP_InertialSensor_Backend::_notify_new_gyro_raw_sample(uint8_t instance,
             _imu._gyro_notch_filter[instance].reset();
             _imu._gyro_harmonic_notch_filter[instance].reset();
         }
+        else {
```
tiny formatting thing but this line should be "} else {"

andyp1per replied:
Fixed, thanks

rmackay9 commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -500,14 +506,15 @@ void AP_InertialSensor_Backend::update_gyro(uint8_t instance)
 
     // possibly update filter frequency
-    if (_last_gyro_filter_hz != _gyro_filter_cutoff()) {
+    if (_last_gyro_filter_hz != _gyro_filter_cutoff() || sensors_converging()) {
```
just checking that the intention here is that the lines below (set_cutoff_frequency, etc) are run all the time during the first 30 seconds that the board is powered up?

andyp1per replied:
@rmackay9 see my comment https://github.com/ArduPilot/ardupilot/pull/11795#issuecomment-526485212 - during this "warmup" phase we can change the sample frequency - this is designed to catch that. Whether this is the right way to warmup is another question but it works and is consistent with the way we are changing the sample frequency.

rmackay9 commented on libraries/Filter/HarmonicNotchFilter.cpp:
```diff
@@ -164,7 +182,7 @@ T HarmonicNotchFilter<T>::apply(const T &sample)
 
     T output = sample;
-    for (uint8_t i = 0; i < _num_filters; i++) {
+    for (uint8_t i = 0; i < _num_enabled_filters; i++) {
```
I wonder if it is possible that the _num_enabled_filters is perhaps 2 but it's actually the 1st and 3rd filters that are enabled.  I.e. are we sure that it will always be the final filters that are not enabled rather than a filter in the middle?

andyp1per replied:
@rmackay9 we only disable filters if they have a cutoff frequency above the nyquist limit. We only _create_ filters if they are in the harmonics bitmask - so in your example there would only be 2 filters created and whether they were enabled or not would depend on the center frequency. The center frequency of each filter is always higher than the previous so we could never enable the 3rd but have the 1st disabled.

rmackay9 commented:
I've lightly tested this with valgrind in SITL.  The test was quick though, I simply set INS_NOTCH_ENABLE = 1 and INS_HNTCH_ENABLE = 1, rebooted and flew the vehicle in stabilize in Loiter.  If there are other setting changes that would be more useful for a valgrind test please reply here.  Just FYI - the point of the valgrind test is to look for memory access errors.  So things like attempts to access memory past the end of an array.

rmackay9 commented:
@tridge, I think this is your area so perhaps you should give the final "OK" on whether you're happy for this to go in now.  Any other inputs from others is greatly appreciated.

andyp1per commented:
> 
> 
> I see the HarmonicNotchFilter is created as a template but is there any plan to make it work with anything besides a Vector3f? If "no" then it could be slightly simplified by removing the template part. This is not necessarily a blocker but just a question.

Ideally we would track the noise on each axis independently and run a separate filter on each - so I think the template part is worth keeping, it also makes it consistent with the rest of the filters

andyp1per commented:
> 
> 
> I've lightly tested this with valgrind in SITL. The test was quick though, I simply set INS_NOTCH_ENABLE = 1 and INS_HNTCH_ENABLE = 1, rebooted and flew the vehicle in stabilize in Loiter. If there are other setting changes that would be more useful for a valgrind test please reply here. Just FYI - the point of the valgrind test is to look for memory access errors. So things like attempts to access memory past the end of an array.

Another test you could do is set the harmonics to 7 (3 filters enabled) and set the FREQ to a quarter of the gyro sample rate (not sure what that is in SITL)

lthall commented:
> I see the HarmonicNotchFilter is created as a template but is there any plan to make it work with anything besides a Vector3f? If "no" then it could be slightly simplified by removing the template part. This is not necessarily a blocker but just a question.

It may be used on an individual axis in the future.

bnsgeyer commented:
@andyp1per  and @lthall  why haven't there been any notch filters on accelerometer sensors.  It's not so much for x and y but I think it would be very helpful in z as this affects the altitude hold controller.  I know with helicopters, I think this would be a big benefit.

andyp1per commented:
@bnsgeyer https://github.com/ArduPilot/ardupilot/pull/11515 - I'm sure @guglie would like some support for getting this in :)

lthall commented:
This is where it may be more efficient to filter just on the axis being used rather than all three axis. 

But we need to ensure that the benefit is worth the cost.

bnsgeyer commented:
@tridge  Here are the logs for the CUAV v5 controller.  I flew one flight using only the single notch filter at 25 hz.  Then I flew the second flight with the harmonic notch to tackle the 50 and 100 hz frequencies.  Let me know if you can't get to the files.  [log files](https://drive.google.com/open?id=1ZMlp73RbTCF5MNFjNcRUSl0IboaZI3aK)

tridge commented:
tested build on fmuv2, and extra flash cost is under 1k, which I think is low enough to be included.
On CPU cost, on a F427 the cost is 0.7% of total CPU per IMU. So 2.1% of the total CPU for a 3 IMU system (ie. it costs 21ms of CPU per second with 3 IMUs).
Do we need to calculate this filter on IMUs other than the one being actively used in the rate controllers? We could save quite a lot of CPU if we only ran it on the active IMU.

andyp1per commented:
@tridge - this really applies to all of the gyro filtering (i.e. low pass and static notch as well) in my mind, the EKF is not using it, so you could simply bypass the gyro filtering on all the unused IMU's. I suppose it does beg the question of whether there's any latency issue when you switch. But you could have a gyro imu mask to control? That said even on my racer I like seeing all of the output as I can then compare.

tridge commented:
@andyp1per I'll leave it up to you and @lthall. 2% isn't a huge cost, but it also isn't trivial. I see the benefit of logs showing the impact on all gyros.
I guess we could add an IMU mask for it?

andyp1per commented:
@tridge thanks! Side question - what procedure do you use to test the CPU load? I'd be happy to document on the wiki.
